init/main.go

```diff
@@ -697,6 +705,57 @@ func checkIfInitrd() error {
 	return nil
 }
 
+func blkInfoMatchesArg(info *blkInfo, arg string) bool {
+	if strings.HasPrefix(arg, "UUID=") && info.uuid == strings.TrimPrefix(arg, "UUID=") {
+		return true
+	}
+	if strings.HasPrefix(arg, "LABEL=") && info.label == strings.TrimPrefix(arg, "LABEL=") {
+		return true
+	}
+
+	return false
+}
+
+func resume(devpath string) error {
+	resume, ok := cmdline["resume"]
+	if !ok {
+		return nil
+	}
+
+	var resumeDevice string
+	if resume == devpath {
+		resumeDevice = devpath
+		goto resume
+	}
+
+	{
+		info, err := readBlkInfo(devpath)
+		if err != nil {
+			return fmt.Errorf("%s: %v", devpath, err)
+		}
+
+		if blkInfoMatchesArg(info, devpath) {
+			resumeDevice = devpath
+			goto resume
+		}
+	}
+
+	return nil
+resume:
+	major, minor, err := deviceNo(resumeDevice)
+	if err != nil {
+		return err
+	}
+
+	rd := fmt.Sprintf("%d:%d", major, minor)
+	debug("resume matched: %s=%s, writing %s", resume, resumeDevice, rd)
+	if err := os.WriteFile("/sys/power/resume", []byte(rd), 0644); err != nil {
```

Comment:
where I can find more information about this "echo $diskno > /sys/power/resume" operation. What exactly kernel does here? How does suspend-to-XXX implemented in Linux?

Reply:
A couple resources:

* https://wiki.archlinux.org/index.php/Power_management/Suspend_and_hibernate#Required_kernel_parameters has a summary of what's going on here

* From the [kernel docs](https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html):
   >>> 
        resume=         [SWSUSP]
                        Specify the partition device for software suspend
                        Format:
                        {/dev/<dev> | PARTUUID=<uuid> | <int>:<int> | <hex>}

        resume_offset=  [SWSUSP]
                        Specify the offset from the beginning of the partition
                        given by "resume=" at which the swap header is located,
                        in <PAGE_SIZE> units (needed only for swap files).
                        See  Documentation/power/swsusp-and-swap-files.rst

* for reference, [mkinitcpio](https://github.com/archlinux/mkinitcpio/blob/master/hooks/resume#L25-L32) has similar logic, as does [dracut](https://github.com/dracutdevs/dracut/blob/master/modules.d/95resume/resume.sh#L17)

One thing that I noticed when implementing this is that it appears that don't need to handle the `resume.offset` kernel parameter in the initrd, just the `resume` parameter - the kernel seems to parse that from its commandline itself when resuming, but it needs `/sys/power/resume` to have the major/minor of the block device set already for the resume to occur.